File: pingdart-js/src/services/ai.service.js
import axios from 'axios';

class AiService {
    constructor(http, config, realtimeHttp) {
        this.http = http;
        // Using the main HTTP client to ensure API Key is attached if present
    }

    async callAiApi(message, onProgress, options = {}) {
        try {
            const payload = {
                message: message,
                stream: true,
                model: options.model || "chinnuai:1.1",
                chinuai: options.chinuai || "chinnuai:1.1",
                user_id: options.userId || "default", // Can be overridden
                ...options
            };

            // If running in a Node.js environment or standard browser
            if (typeof fetch === 'function' && !this.http.defaults.adapter) { // Checking if we're not strictly forced into axios node adapter
                // Fallback to fetch if available and preferred for streams in some environments
                // But since we use Axios from PingDartSDK, we should use the configured axios instance
                // We'll proceed to the Axios route below as primary since it has the x-api-key setup
            }

            // Using the configured Axios client (this.http) ensuring headers (`x-api-key`) are passed
            const response = await this.http({
                method: 'post',
                url: '/api/ai/chinuai-chat',
                data: payload,
                responseType: 'stream', // Important for node.js axios streaming
                // if run in browser, axios responseType: 'stream' might not work directly, but since this is Node.js SDK, it's fine.
            });

            return new Promise((resolve, reject) => {
                let fullResult = "";

                // Node.js Axios stream handling
                if (response.data && typeof response.data.on === 'function') {
                    response.data.on('data', (chunk) => {
                        try {
                            // The backend sends Server-Sent Events (SSE) format: 'data: {"chunk": "...", "done": false}\n\n'
                            const lines = chunk.toString().split('\n').filter(line => line.trim());

                            for (const line of lines) {
                                if (line.startsWith('data: ')) {
                                    const jsonStr = line.replace('data: ', '').trim();
                                    if (!jsonStr) continue;

                                    const parsed = JSON.parse(jsonStr);

                                    if (parsed.chunk) {
                                        fullResult += parsed.chunk;
                                        if (onProgress) onProgress(parsed.chunk);
                                    }

                                    if (parsed.done) {
                                        resolve(fullResult);
                                    }
                                }
                            }
                        } catch (e) {
                            // Sometimes chunks might be fragmented, standard error handling might misinterpret
                            // For raw text streams fallback:
                            const strChunk = chunk.toString();
                            if (!strChunk.includes('data: {')) {
                                fullResult += strChunk;
                                if (onProgress) onProgress(strChunk);
                            }
                        }
                    });

                    response.data.on('end', () => resolve(fullResult));
                    response.data.on('error', reject);
                } else if (response.data) {
                    // Fallback if not a true stream (e.g. browser polyfill or stream config failed)
                    // If the backend sent a full JSON response instead
                    if (response.data.reply) {
                        const replyStr = typeof response.data.reply === 'string' ? response.data.reply : JSON.stringify(response.data.reply);
                        if (onProgress) onProgress(replyStr);
                        resolve(replyStr);
                    } else {
                        resolve(response.data);
                    }
                } else {
                    resolve(fullResult);
                }
            });

        } catch (error) {
            throw new Error(`AI API Error: ${error.message}`);
        }
    }
}

export default AiService;

File: pingdart-python/pingdart_sdk/services/ai_service.py
import requests
import json

class AiService:
    def __init__(self, http_client):
        self.http = http_client

    def call_ai_api(self, message, on_progress=None, model="chinnuai:1.1", **kwargs):
        """
        Call the AI API with optional streaming progress.
        """
        payload = {
            "message": message,
            "stream": True,
            "model": model,
            **kwargs
        }
        
        try:
            # We use the base http client but for streaming we need to be careful with its configuration
            # Assuming self.http is a requests.Session or similar
            base_url = self.http.base_url if hasattr(self.http, 'base_url') else ""
            headers = self.http.headers
            
            response = requests.post(
                f"{base_url}/api/ai/chinuai-chat",
                json=payload,
                headers=headers,
                stream=True
            )
            response.raise_for_status()
            
            full_result = ""
            for line in response.iter_lines():
                if line:
                    decoded_line = line.decode('utf-8')
                    if decoded_line.startswith('data: '):
                        data_str = decoded_line[6:].strip()
                        if not data_str:
                            continue
                        try:
                            parsed = json.loads(data_str)
                            chunk = parsed.get('chunk', '')
                            if chunk:
                                full_result += chunk
                                if on_progress:
                                    on_progress(chunk)
                            if parsed.get('done'):
                                break
                        except json.JSONDecodeError:
                            # Fallback if the line is raw text
                            full_result += decoded_line
                            if on_progress:
                                on_progress(decoded_line)
            
            return full_result
        except Exception as e:
            raise Exception(f"AI API Error: {str(e)}")
